### ISSUE TITLE
listenTo change in behavior, 1.2.0?

### ISSUE THREAD

**Issue by bazineta** [jashkenas/backbone#3611]
Using the 1.1.2 release, listenTo() was able to listen for events from objects that provide the typical on() interface; in our case, an Ace editor session, which emits 'change' events, e.g., EditSession.on('change', function (e)).  Under 1.1.2, we've used listenTo() to listen for these events from the editor session.

Using the 1.2.0 release, listenTo() does not appear to be able to listen for these events; they're emitted, and the editor session is present in the View's _listeningTo, but the callback is no longer fired.

We were using listenTo() for automatic cleanup of the listener function.  Curious if this was never supposed to work in this manner, or if it's a bug.  The changes to listening in 1.2.0 are quite extensive, so it's a bit difficult to see what might be the key difference for one unfamiliar with them.

**Comment by akre54** [comment]
Mind writing up a small test case so we can take a look? I'd be interested to see what your code looks like.

**Comment by jridgewell** [comment]
```javascript
// Test Case
var events = {};
var obj = {
  // Any non Backbone.Events event implementation will work...
  on(name, callback) {
    events[name] = callback;
  },
  off(name) {
    events[name] = null;
  }
};

var view = new Backbone.View();
view.listenTo(obj, 'test', console.log);
view.stopListening();
```

This is a side-effect of fixing https://github.com/jashkenas/backbone/pull/3455#issuecomment-71677970. Since now both the listener and the listenee should be able to clean up memory in the other object, I had to share private state between the two objects. [`internalOn`](https://github.com/jashkenas/backbone/blob/dba0faad09d40818fb85dcf84c1e7b97543d3108/backbone.js#L121) is used so that this private data is never leaked outside of Backbone. But it also means that `#listenTo` uses only internal Backbone code instead of the public `#on`.

I'm able to [fix](https://github.com/jashkenas/backbone/compare/master...jridgewell:private-listening-public-on) this by sharing the state privately inside Backbone instead of through a private function. I'll have to look into the performance impact.

**Comment by akre54** [comment]
Ah. I'd say it's not in our contract to support other events systems, but if we can make it work we probably should. Is there any way other than a singleton event @jridgewell? This seems pretty hacky and brittle.

**Comment by jridgewell** [comment]
Reverting https://github.com/jashkenas/backbone/pull/3594 should also do the trick, since the listener will keep track of the events.

**Comment by akre54** [comment]
What did #3455 and #3594 get us? Kind of hard to follow those discussions.

**Comment by jashkenas** [comment]
Since, #3594 is your baby, @jridgewell — go ahead and revert it if you think that it's reasonable and the appropriate fix.

**Comment by jridgewell** [comment]
#3455 fixed a pretty sever memory leak:

```javascript
obj.on('event', function() {});
obj2.listenTo(obj, 'event', function() {});

obj2.stopListening();
// obj2 still holds a reference to obj
```

It also allowed the listenee can clean up references:

```javascript
obj2.listenTo(obj, 'event', function() {});

obj.off();
// obj2 still holds a reference to obj
```

- - -

Let me get test cases setup, and I'll revert once I know it'll fix the library interop.

**Comment by jridgewell** [comment]
Oh man, even if we revert #3594, the previous implementation used a private `internalStopListening` that doesn't call the public `#off`. Reverting won't fix.

I'll look into what we can do to resolve this, but it's not going to be a pretty solution.

**Comment by jashkenas** [comment]
Point of clarification — the current approach is fine, except that you can't `.on(` to an event from a different events library? Yes?

**Comment by jridgewell** [comment]
> the current approach is fine, except that you can't .on( to an event from a different events library? Yes?

Exactly.

**Comment by jashkenas** [comment]
Then let's leave the current approach as it stands. 

@bazineta Is it not a viable approach for your Ace editor session object to mixin Backbone.Events ?

**Comment by bazineta** [comment]
A number of viable approaches exist; that's one of them.  Agree, FWIW, that there don't seem to be good paths forward from here with respect to changing the current behavior, so I'd instead recommend documenting it as a warning, i.e., this would have worked before, does not now, will fail silently, and should be addressed via other means.

**Comment by gaperton** [comment]
@jashkenas "Point of clarification — the current approach is fine, except that you can't .on( to an event from a different events library? Yes?"

No. Message maps are broken. In this code triggering a, b, or c won't fire any events.

```
this.listenTo( this, {
                            'hello' : function(){
                                this.hello = true;
                            },

                            'a b c' : function(){
                                this.abc = true;
                            }
                        });
```

**Comment by stevewillard** [comment]
Having a similar problem to @gaperton with Backbone 1.2.1 and Marionette 2.4.2. The modelEvents mapping object doesn't seem to trigger "change" "sync" or other callbacks. When I revert back to Backbone 1.1.2, everything works.

**Comment by jridgewell** [comment]
> No. Message maps are broken. In this code triggering a, b, or c won't fire event handler.

The normal object syntax works fine, only the space-separated object syntax has a bug: https://github.com/jashkenas/backbone/pull/3680

**Comment by stevewillard** [comment]
Ah, that explains things. Thanks! I'll be looking for the next Backbone version bump :)

**Comment by akre54** [comment]
Was that actually being used anywhere? It didn't work before so it should've affected anything.

**Comment by stevewillard** [comment]
You're asking if the space separated callback syntax is used anywhere? It's documented in Marionette:

http://marionettejs.com/docs/marionette.view.html#multiple-callbacks

```
Marionette.CompositeView.extend({

  modelEvents: {
    "change:name": "nameChanged thatThing"
  },

  nameChanged: function(){ },

  thatThing: function(){ },
});
```

**Comment by gaperton** [comment]
@stevewillard, in the case you are concerned, you can use this code as workaround.

{code}
var Events = Backbone.Events,
    bbListenTo = Events.listenTo;

Events.listenTo = function( obj, events ){
    if( typeof events === 'object' ){
        for( var event in events ) bbListenTo.call( this, obj, event, events[ event ] );
        return this;
    }

    return bbListenTo.apply( this, arguments );
};

[ 'Model', 'Collection', 'View', 'Router', 'History' ].forEach( function( name ){
    var Type = Backbone[ name ];
    Type.prototype.listenTo = Events.listenTo;
});
{code}

**Comment by akre54** [comment]
I'm not sure I understand. We (Backbone) don't use strings for callbacks anywhere. #3680 deals with the keys of the listeners object (the event names), not the values (the callbacks). Having multiple event names in the keys of the listener objects wasn't supported previously so that can't be causing breakage.

As I understand it, Marionette uses `bindEntityEvents` under the hood to [split strings and `listenTo` each callback separately](https://github.com/marionettejs/backbone.marionette/blob/d182dc3f09b13a2603c142029d45994e1455105c/src/bind-entity-events.js#L21-L36).

> @akre54, yes, it is used somewhere. Otherwise no one would made comments about it.

Can you point out the specific line? Telling me it broke something doesn't help unless we can track it down.

**Comment by gaperton** [comment]
@akre54, when I upgraded backbone version to this one, unit tests in my framework started to fail. This one, if you're interested.

https://github.com/Volicon/backbone.nestedTypes

I figured out why, and posted quite a clean example to the comments above about a month ago.

If you think that it's too big deal to fix it, it's fine. Anyway I've got a lot of backbone things overridden, one more, one less, it doesn't matter. Just though you might be concerned.

**Comment by akre54** [comment]
> I figured out why, and posted quite a clean example to the comments above about a month ago.

You're referring to https://github.com/jashkenas/backbone/issues/3611#issuecomment-104882343? That second form has never worked. We literally just added support for it [this morning](https://github.com/jashkenas/backbone/pull/3680). 

Can you point to a line in your test code where this is broken? I can't see anything obvious in that repository you linked and I don't have time to follow down rabbit holes.

**Comment by akre54** [comment]
Sorry -- my mistake. It looks like the `obj.on({'a b c': fn})` form was working in 1.1.2. #3680 should fix this and will be out shortly.

**Comment by gaperton** [comment]
> We literally just added support for it this morning.

Nice to hear that.

> Can you point to a line in your test code where this is broken?

Yes, if you insist. But it will not be easy :)

This test and some other tests related to nested Collection change detection were broken.
https://github.com/Volicon/backbone.nestedTypes/blob/master/test/backboneTypes.js#L188

That's because Collection has several reasons to be changed.
https://github.com/Volicon/backbone.nestedTypes/blob/master/src/collection.js#L25

Internally, in Model/Collection metatype definition, message map is being used for subscription.
https://github.com/Volicon/backbone.nestedTypes/blob/master/src/metatypes.js#L181

Which, in turn created the bug. Events are ignored, no nested collection changes are detected. Of course, since I cannot live with such a bad thing, I spend a couple of minutes adding few lines of code here:
https://github.com/Volicon/backbone.nestedTypes/blob/master/src/backbone%2B.js#L18

And, trying to be a good boy, reported the bug to maintainers. That's all the story :)

**Comment by jridgewell** [comment]
Moving to #3615.
